### PR TITLE
Unit testing order

### LINKED ISSUE #8117: Explore combining the different language topics for .NET Core unit test
We have one entry per language but this process is painful when we need to make updates. We should look into this and see if it makes sense to combine them into one per technology and have tabs for each language.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: a64f7e75-fbea-18ba-f290-3e5bb3ab0da7
* Version Independent ID: 24639c20-6c3a-738e-2ff8-9a6ee6f232dc
* Content: [Unit Testing in .NET Core](https://docs.microsoft.com/en-us/dotnet/core/testing/)
* Content Source: [docs/core/testing/index.md](https://github.com/dotnet/docs/blob/master/docs/core/testing/index.md)
* Product: **dotnet-core**
* GitHub Login: @ardalis
* Microsoft Alias: **wiwagn**

### LINKED ISSUE #16129: Update VB unit testing with xUnit
Update to follow [Unit testing C# in .NET Core using dotnet test and xUnit](https://docs.microsoft.com/en-us/dotnet/core/testing/unit-testing-with-dotnet-test) in PR #15954

image

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 39c0da0e-ebb1-8949-22e6-e66eadf05d41
* Version Independent ID: 809a1bd4-28a6-ca05-d5e5-b27dae4c5c61
* Content: [Unit testing Visual Basic in .NET Core with dotnet test and xUnit - .NET Core](https://docs.microsoft.com/en-us/dotnet/core/testing/unit-testing-visual-basic-with-dotnet-test)
* Content Source: [docs/core/testing/unit-testing-visual-basic-with-dotnet-test.md](https://github.com/dotnet/docs/blob/master/docs/core/testing/unit-testing-visual-basic-with-dotnet-test.md)
* Product: **dotnet-core**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

### LINKED ISSUE #18432: Missing documentation on how to run .NET Core unit tests sequentially
I want to run unit tests sequentially in .net core but could not find any solutions. The solutions available on the Internet do not work for me. I would expect a detailed discussion on this topic.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: bb70526f-d02c-2b6c-3d34-0f4cac5809f0
* Version Independent ID: e4443d71-7131-ae9b-447d-9ac0762a99fc
* Content: [Running selective unit tests - .NET Core](https://docs.microsoft.com/en-us/dotnet/core/testing/selective-unit-tests#feedback)
* Content Source: [docs/core/testing/selective-unit-tests.md](https://github.com/dotnet/docs/blob/master/docs/core/testing/selective-unit-tests.md)
* Product: **dotnet-core**
* GitHub Login: @smadala
* Microsoft Alias: **wiwagn**

### PR DESCRIPTION
**Pull request by IEvangelist**
## Summary

- Wrote new article "Order unit tests" for MSTest, xUnit, and NUnit
- Remove duplicate TOC entry, "Unit testing" and "Overview" were the same
- Add `xref` where possible in "Selective unit tests" doc
- Allow TOC to scope to test framework
- Add `dotnet/samples` repo to *.openpublishing.publish.config.json*
- Corrected redirect link, https://xunit.net/
- Add zone pivots def to *docs/zone-pivot-groups.yml*
- Use zone pivots for different test frameworks
  - Overview, Run select unit tests, and Order unit tests
- Break out long `dotnet test --filter` commands into code blocks for readability
- Normalize *docfx.json* file
- Fixed heading that use "ing"
- Add `dotnet *` links where missing
- Add `displayName` to help with filtering
- Rewrite VB article per #16129

Fixes #18432, fixes #16129, closes #8117 (as this was explored as part of this effort)

### Preview links

✔️ [Unit testing overview](https://review.docs.microsoft.com/en-us/dotnet/core/testing/?branch=pr-en-us-18456)
✔️ [Run selective unit tests](https://review.docs.microsoft.com/en-us/dotnet/core/testing/selective-unit-tests?branch=pr-en-us-18456)
✔️ [Order unit tests](https://review.docs.microsoft.com/en-us/dotnet/core/testing/order-unit-tests?branch=pr-en-us-18456)

### CODE REVIEW

File: docs/core/toc.yml [hunk 1]
```diff
@@ -365,32 +365,47 @@
...
+      href: testing/unit-testing-with-nunit.md
+    - name: F# unit testing
+      href: testing/unit-testing-fsharp-with-nunit.md
+    - name: VB unit testing
+      href: testing/unit-testing-visual-basic-with-nunit.md
+    - name: Run selective unit tests
+      href: testing/selective-unit-tests.md?tab=nunit
+    - name: Order unit tests
+      href: testing/order-unit-tests.md?tab=nunit
+  - name: MSTest
```
Comment (IEvangelist) [inline docs/core/toc.yml:398]:
```suggestion
  - name: MSTest
    items:
```

File: docs/core/testing/order-unit-tests.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+Then in a test class you would set the test case order with the `TestCaseOrdererAttribute` to the `PriorityOrderer`.
+
+:::code language="csharp" source="~/dotnet-samples/csharp/unit-testing/XUnit.TestProject/ByPriorityOrder.cs":::
+
+## [NUnit](#tab/nunit)
+
+### Order alphabetically
+
+With NUnit, tests are automatically ordered by their test name.
+
```
Comment (ChrisMaddock) [inline docs/core/testing/order-unit-tests.md:71]: Nice article! A correction here - this is actually considered an implementation detail of NUnit, and could change in future. See here: https://github.com/nunit/nunit/issues/1372

Strictly speaking, NUnit has no defined default ordering - the order attribute is the way to enforce it if needed. 🙂

File: docs/core/testing/order-unit-tests.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+### Order by priority
+
+To order tests explicitly, NUnit provides an [`OrderAttribute`](https://github.com/nunit/docs/wiki/Order-Attribute). Tests with this attribute are started before tests without. The order value is used to determined the order to run the unit tests.
+
+:::code language="csharp" source="~/dotnet-samples/csharp/unit-testing/NUnit.TestProject/ByOrder.cs":::
+
+---
+
+## Nest Steps
```
Comment (ChrisMaddock) [inline docs/core/testing/order-unit-tests.md:85]:
```suggestion
## Next Steps
```

🙂

File: docs/core/testing/order-unit-tests.md (hunk 2 above)
Comment (IEvangelist) [reply docs/core/testing/order-unit-tests.md]: Thank you very much @ChrisMaddock - this information is very helpful. If you and the NUnit team decide to add this functionality later or, I'd be happy to collaborate with you on adding it. 🙏

File: docs/core/testing/order-unit-tests.md (hunk 3 above)
Comment (IEvangelist) [reply docs/core/testing/order-unit-tests.md]: Oops, nice catch.

File: docs/core/testing/selective-unit-tests.md [hunk 4]
```diff
@@ -1,27 +1,34 @@
...
-Note the backslash that precedes the exclamation mark.
+Using filters that include exclamation mark `!` on `*nix` requires escaping since `!` is reserved. For example, this filter
+skips all tests if the namespace contains IntegrationTests:
+
+```dotnetcli
+dotnet test --filter FullyQualifiedName\!~IntegrationTests
+```
+
+> [!IMPORTANT]
+> Notice the backslash that precedes the exclamation mark, that indicates it is an escaped character `\!`.
```
Comment (BillWagner) [inline docs/core/testing/selective-unit-tests.md:23]: suggest rewording to avoid *that* repeated:

```suggestion
> The backslash precedes the exclamation mark to indicate it is an escaped character `\!`.
```

File: docs/core/testing/unit-testing-visual-basic-with-dotnet-test.md [hunk 5]
```diff
@@ -3,97 +3,139 @@ title: Unit testing Visual Basic in .NET Core with dotnet test and xUnit
...
 author: billwagner
 ms.author: wiwagn
-ms.date: 09/01/2017
+ms.date: 05/18/2020
 ---
 # Unit testing Visual Basic .NET Core libraries using dotnet test and xUnit
 
-This tutorial takes you through an interactive experience building a sample solution step-by-step to learn unit testing concepts. If you prefer to follow the tutorial using a pre-built solution, [view or download the sample code](https://github.com/dotnet/samples/tree/master/core/getting-started/unit-testing-vb-dotnet-test) before you begin. For download instructions, see [Samples and Tutorials](../../samples-and-tutorials/index.md#viewing-and-downloading-samples).
+This tutorial shows how to build a solution containing a unit test project and source code project. To follow the tutorial using a pre-built solution, [view or download the sample code](https://github.com/dotnet/samples/tree/master/core/getting-started/unit-testing-using-dotnet-test/). For download instructions, see [Samples and Tutorials](../../samples-and-tutorials/index.md#viewing-and-downloading-samples).
```
Comment (BillWagner) [inline docs/core/testing/unit-testing-visual-basic-with-dotnet-test.md:10]: nit: I'd say library project. (Both projects have source code):

```suggestion
This tutorial shows how to build a solution containing a unit test project and library project. To follow the tutorial using a pre-built solution, [view or download the sample code](https://github.com/dotnet/samples/tree/master/core/getting-started/unit-testing-using-dotnet-test/). For download instructions, see [Samples and Tutorials](../../samples-and-tutorials/index.md#viewing-and-downloading-samples).
```

**Review by BillWagner — APPROVED** [review]
Overall this looks great @IEvangelist 

I had some sentence structure suggestions, and one question on the redirection.

Once ready, you can :shipit:

File: .openpublishing.publish.config.json [hunk 6]
```diff
@@ -74,6 +74,11 @@
       "path_to_root": "machinelearning-samples",
       "url": "https://github.com/dotnet/machinelearning-samples",
       "branch": "master"
+    },
+    {
```
[the remark this answers is not in the document]
Comment (IEvangelist) [reply .openpublishing.publish.config.json:78]: This enables samples to be rendered as snippets:

```markdown
:::code language="csharp" source="~/dotnet-samples/csharp/unit-testing/NUnit.TestProject/ByOrder.cs":::
```

I could remove this entry from the *.openpublishing.publish.config.json*, then I'd have to duplicate all the source code and update all of these snippets to point locally to the duplicated samples. I'll let you decide what you prefer.

[the remark this answers is not in the document]
Comment (BillWagner) [reply .openpublishing.publish.config.json:78]: I didn't know this was possible.

Can you start a discussion with the team? I'd like others to weigh in before we standardize. (Especially @Thraka )

In the meantime, I'm fine with merging this in its current form. Whatever we decide, we'll have changes to make....

[the remark this answers is not in the document]
Comment (IEvangelist) [reply .openpublishing.publish.config.json:78]: It's actually part of the tooling and fully supported by the platform. For example, when you use the <kbd>Alt+M</kbd> while in a markdown file, and select snippet - the tooling lights up to rely on this repo and inserts the snippet. We used it all the time in azure-docs. I'll start a conversation. Thanks
